src/Compilers/Core/Portable/PEWriter/CustomDebugInfoWriter.cs

```diff
@@ -123,6 +103,27 @@ public bool ShouldForwardNamespaceScopes(EmitContext context, IMethodBody method
 
             byte[] result = encoder.ToArray() ?? Array.Empty<byte>();
             pooledBuilder.Free();
+
+            emitExternNamespaces = false;
+
+            // CONSIDER: this may not be the same "first" method as in Dev10, but
+            // it shouldn't matter since all methods will still forward to a method
+            // containing the appropriate information.
+            if (_methodBodyWithModuleInfo == null && result.Length > 0)
```

Comment:
>result.Length > 0 [](http://example.com/codeflow?start=53&length=17)

It looks like this condition makes a difference at the call site only when ```emitAllDebugInfo``` is false, and, it looks like in that case we are not interested in emitting information about extern aliases even if the blob was not empty. Here we are applying the filter unconditionally. This feels fragile to me, it is not obvious we are not going to break something. If our goal is to skip synthesized methods, perhaps we can filter those out explicitly here, or in the caller?

Reply:
The `Length` check is to ensure we're only notifying the caller to emit extern aliases if we've generated some other debug info here. The goal is to skip methods where no debug info is needed, not necessarily synthesized methods.

Reply:
Good catch, this is more fragile than I'd thought. Changed to check for `emitStateMachineInfo` directly and added an assert in the caller.